src/test/java/com/microsoft/sqlserver/jdbc/datatypes/KatmaiDataTypesTest.java

```diff
@@ -0,0 +1,1684 @@
+
+                    // Scale is interpreted as number of fractional seconds precision
+                    assertEquals(metadata.getScale(1), scale, "getScale() of " + sqlCastExpression());
+                    assertEquals(metadata.getParameterClassName(1), sqlType.className,
+                            "getParameterClassName() of " + sqlCastExpression());
+                    // Katmai temporal types are not signed
+                    assertEquals(metadata.isSigned(1), false, "isSigned() of " + sqlCastExpression());
+                }
+            } finally {
+                try (Statement stmt = conn.createStatement()) {
+                    stmt.executeUpdate("DROP PROCEDURE " + AbstractSQLGenerator.escapeIdentifier(procName));
+                }
+            }
+        }
+
+        abstract void verifyRSGetters(ResultSet rs) throws Exception;
+
+        void verifyRSGetters(Connection conn) throws Exception {
+            try (Statement stmt = conn.createStatement(); ResultSet rs = stmt.executeQuery(
+                    "SELECT " + sqlCastExpression() + ", CAST(" + sqlCastExpression() + " AS VARCHAR(60))")) {
+                rs.next();
+                verifyRSGetters(rs);
+            }
+        }
+
+        abstract void verifyRSUpdaters(ResultSet rs) throws Exception;
+
+        void verifyRSUpdaters(Connection conn) throws Exception {
+
+            assumeTrue(!TestUtils.isSqlAzureDW(conn), TestResource.getResource("R_skipAzure"));
+
+            try (Statement stmt = conn.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.CONCUR_UPDATABLE)) {
+
+                stmt.executeUpdate("DROP TABLE " + AbstractSQLGenerator.escapeIdentifier(tableName));
+
+                stmt.executeUpdate("CREATE TABLE " + AbstractSQLGenerator.escapeIdentifier(tableName) + " (col1 "
+                        + sqlTypeExpression + ", col2 int identity(1,1) primary key)");
+
+                stmt.executeUpdate("INSERT INTO " + AbstractSQLGenerator.escapeIdentifier(tableName) + " VALUES ("
+                        + sqlCastExpression() + ")");
+
+                try (ResultSet rs = stmt
+                        .executeQuery("SELECT * FROM " + AbstractSQLGenerator.escapeIdentifier(tableName))) {
+                    rs.next();
+                    verifyRSUpdaters(rs);
+                } finally {
+                    stmt.executeUpdate("DROP TABLE " + AbstractSQLGenerator.escapeIdentifier(tableName));
+                }
+            }
+        }
+
+        /*
+         * For testing the setObject and setNull methods in PreparedStatement, use the verifySetter* methods. These
+         * methods prepare a single statement and execute it for all different data types by calling the appropriate
+         * 'setObject' methods for each data type and/or type conversion.
+         */
+        abstract void verifySetters(PreparedStatement ps) throws Exception;

```

Comment:
Please move all abstract methods to the top of the class with Javadocs instead of comment.